core/src/main/scala/org/http4s/HeaderV2.scala

```diff
@@ -0,0 +1,333 @@
+      new Header.ToRaw {
+        val value = Header.Raw(H.name, H.value(h))
+      }
+  }
+
+  /** Abstracts over Single and Recurring Headers
+    */
+  sealed trait Select[A] {
+    type F[_]
+
+    /** Transform this header into a [[Header.Raw]]
+      */
+    def toRaw(a: A): Header.Raw
+
+    /** Selects this header from a list of [[Header.Raw]]
+      */
+    def from(headers: List[Header.Raw]): Option[F[A]]
+  }
+  trait LowPrio {
+    implicit def recurringHeadersNoMerge[A](implicit
+        h: Header[A, Header.Recurring]): Select[A] { type F[B] = NonEmptyList[B] } =
+      new Select[A] {
+        type F[B] = NonEmptyList[B]
+
+        def toRaw(a: A): Header.Raw =
+          Header.Raw(h.name, h.value(a))
+
+        def from(headers: List[Header.Raw]): Option[NonEmptyList[A]] =
+          headers.collect(Function.unlift(Select.fromRaw(_))).toNel
+      }
+  }
+  object Select extends LowPrio {
+    def fromRaw[A](h: Header.Raw)(implicit ev: Header[A, _]): Option[A] =
+      (h.name == Header[A].name).guard[Option] >> Header[A].parse(h.value)
+
+    implicit def singleHeaders[A](implicit
+        h: Header[A, Header.Single]): Select[A] { type F[B] = Id[B] } =
+      new Select[A] {
+        type F[B] = Id[B]
+
+        def toRaw(a: A): Header.Raw =
+          Header.Raw(h.name, h.value(a), recurring = false)
+
+        def from(headers: List[Header.Raw]): Option[A] =
+          headers.collectFirst(Function.unlift(fromRaw(_)))
+      }
+
+    implicit def recurringHeadersWithMerge[A: Semigroup](implicit
+        h: Header[A, Header.Recurring]): Select[A] { type F[B] = Id[B] } =
+      new Select[A] {
+        type F[B] = Id[B]
+
+        def toRaw(a: A): Header.Raw =
+          Header.Raw(h.name, h.value(a))
+
+        def from(headers: List[Header.Raw]): Option[A] =
+          headers.foldLeft(Option.empty[A]) { (a, raw) =>
+            fromRaw(raw) match {
+              case Some(aa) => a |+| aa.some
+              case None => a
+            }
+          }
+      }
+  }
+}
+
+/** A collection of HTTP Headers */
+final class Headers(val headers: List[Header.Raw]) extends AnyVal {
+
+  /** TODO revise scaladoc
+    * Attempt to get a [[org.http4s.Header]] of type key.HeaderT from this collection
+    *
+    * @param key [[HeaderKey.Extractable]] that can identify the required header
+    * @return a scala.Option possibly containing the resulting header of type key.HeaderT
+    * @see [[Header]] object and get([[org.typelevel.ci.CIString]])
+    */
+  def get[A](implicit ev: Header.Select[A]): Option[ev.F[A]] =
+    ev.from(headers)
+
+  /** TODO revise scaladoc
+    * Attempt to get a [[org.http4s.Header]] from this collection of headers
+    *
+    * @param key name of the header to find
+    * @return a scala.Option possibly containing the resulting [[org.http4s.Header]]
+    */
+  def get(key: CIString): Option[NonEmptyList[Header.Raw]] = headers.filter(_.name == key).toNel
```

Comment:
Would it be simpler to just return `List[Header.Raw]`?

Reply:
For some reason I like `Option[NonEmptyList` here, I think the emphasis is more on the fact that the header might not be present, and I'd imagine you'd want to do a bunch of `getOrElse`, whereas it's not as easy to do `onEmpty` on `List`. Certainly up for debate though. Also the typed `get` returns `Option[NonEmptyList[A]]` for recurring headers, so there's some consistency there